rabbitsmpp/multiconsumer.go

```diff
@@ -0,0 +1,167 @@
+package rabbitsmpp
+
+import (
+	"errors"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	errConsumerDuplicate = errors.New("consumer already exist")
+	errConsumerMissing   = errors.New("consumer missing")
+)
+
+// ConsumerContainer is a container that collects the data from all the consumer channels into 1
+// channel it's a fan-in pattern. You can add an remove consumers dynamically
+type ConsumerContainer struct {
+	consumers  map[string]Consumer
+	m          *sync.Mutex
+	wg         *sync.WaitGroup
+	Ctx        context.Context
+	stopper    context.CancelFunc
+	outJobChan chan Job
+	outErrChan chan error
+}
+
+func NewContainer() *ConsumerContainer {
+	consumers := make(map[string]Consumer)
+	var m sync.Mutex
+	var wg sync.WaitGroup
+
+	outJobChan := make(chan Job)
+	outErrChan := make(chan error)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &ConsumerContainer{
+		consumers:  consumers,
+		m:          &m,
+		wg:         &wg,
+		Ctx:        ctx,
+		stopper:    cancel,
+		outJobChan: outJobChan,
+		outErrChan: outErrChan,
+	}
+}
+
+func (container *ConsumerContainer) getNoLock(consumerID string) (Consumer, error) {
+	consumer, ok := container.consumers[consumerID]
+	if !ok {
+		return nil, errConsumerMissing
+	}
+
+	return consumer, nil
+}
+
+func (container *ConsumerContainer) removeNoLock(consumerID string) error {
```

Comment:
yeah, I'm a little bit extra careful when locking stuff so, wanted to be extra explicit :)